hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -914,91 +900,86 @@ def lower_common_block_offset(self):
         self.__last_checked_changed_at = self.shrink_target
-        self.__all_changed_blocks = set()
+        self.__all_changed_nodes = set()
 
     def mark_changed(self, i):
-        self.__changed_blocks.add(i)
+        self.__changed_nodes.add(i)
 
     @property
-    def __changed_blocks(self):
-        if self.__last_checked_changed_at is not self.shrink_target:
-            prev_target = self.__last_checked_changed_at
-            new_target = self.shrink_target
-            assert prev_target is not new_target
-            prev = prev_target.buffer
-            new = new_target.buffer
-            assert sort_key(new) < sort_key(prev)
-
-            if (
-                len(new_target.blocks) != len(prev_target.blocks)
-                or new_target.blocks.endpoints != prev_target.blocks.endpoints
-            ):
-                self.__all_changed_blocks = set()
-            else:
-                blocks = new_target.blocks
-
-                # Index of last block whose contents have been modified, found
-                # by checking if the tail past this point has been modified.
-                last_changed = binary_search(
-                    0,
-                    len(blocks),
-                    lambda i: prev[blocks.start(i) :] != new[blocks.start(i) :],
-                )
-
-                # Index of the first block whose contents have been changed,
-                # because we know that this predicate is true for zero (because
-                # the prefix from the start is empty), so the result must be True
-                # for the bytes from the start of this block and False for the
-                # bytes from the end, hence the change is in this block.
-                first_changed = binary_search(
-                    0,
-                    len(blocks),
-                    lambda i: prev[: blocks.start(i)] == new[: blocks.start(i)],
-                )
+    def __changed_nodes(self):
+        if self.__last_checked_changed_at is self.shrink_target:
+            return self.__all_changed_nodes
+
+        prev_target = self.__last_checked_changed_at
+        new_target = self.shrink_target
+        assert prev_target is not new_target
+        prev_nodes = prev_target.examples.ir_tree_nodes
+        new_nodes = new_target.examples.ir_tree_nodes
+        assert sort_key(new_target.buffer) < sort_key(prev_target.buffer)
+
+        if len(prev_nodes) != len(new_nodes) or any(
+            n1.ir_type != n2.ir_type for n1, n2 in zip(prev_nodes, new_nodes)
+        ):
+            # should we check kwargs are equal as well?
+            self.__all_changed_nodes = set()
+        else:
+            assert len(prev_nodes) == len(new_nodes)
+            for i, (n1, n2) in enumerate(zip(prev_nodes, new_nodes)):
+                assert n1.ir_type == n2.ir_type
+                if ir_value_equal(n1.ir_type, n1.value, n2.value):
+                    continue
+                self.__all_changed_nodes.add(i)
```

Comment:
idiom: for simple cases like this, I usually prefer to invert the condition rather than use a `continue`.

Reply:
will try to remember that (and I've changed it here), but you may have to continue reminding me. I personally am used to + prefer the opposite idiom of "list all guard cases then write the logic", to reduce nesting. Even in the one guard case. (but am open to convincing...)

Reply:
🤷‍♂️ either is fine, honestly.  I care more about overall volume and ~complexity of the code and it's really an aesthetic thing.